engine/src/engine.rs

```diff
@@ -280,27 +531,67 @@ impl<'a> ModuleImportResolver for RuntimeModuleImportResolver {
     }
 }
 
-pub fn exec<T: TrackingCopy, G: GlobalState<T>>(
-    module: wasmi::Module,
-    account_addr: [u8; 20],
-    gs: &G,
-) -> Result<ExecutionEffect, Error> {
+fn instance_and_memory(parity_module: Module) -> Result<(ModuleRef, MemoryRef), Error> {
+    let module = wasmi::Module::from_parity_wasm_module(parity_module)?;
     let resolver = RuntimeModuleImportResolver::new();
     let mut imports = ImportsBuilder::new();
     imports.push_resolver("env", &resolver);
     let instance = ModuleInstance::new(&module, &imports)?.assert_no_start();
 
     let memory = resolver.mem_ref()?;
+    Ok((instance, memory))
+}
+
+fn sub_call<T: TrackingCopy>(
+    parity_module: Module,
+    args: Vec<Vec<u8>>,
+    current_runtime: &mut Runtime<T>,
+) -> Result<Vec<u8>, Error> {
+    let (instance, memory) = instance_and_memory(parity_module.clone())?;
+    let known_urefs: HashSet<Key> = HashSet::new();
+    let mut runtime = Runtime {
+        args,
+        memory,
+        state: current_runtime.state,
+        known_urefs,
+        module: parity_module,
+        result: Vec::new(),
+    };
+
+    let result = instance.invoke_export("call", &[], &mut runtime);
+
+    match result {
+        Ok(_) => Ok(runtime.result),
+        Err(e) => {
+            if let Some(host_error) = e.as_host_error() {
+                if let Error::Ret = host_error.downcast_ref::<Error>().unwrap() {
+                    return Ok(runtime.result);
```

Comment:
We only return `Ok` in the case of `Error::Ret`, not all host errors. Other host errors are still captured in `Err`. The reason for this mechanism is to stop execution of the wasm program early because it has returned a value, like the `return` keyword in other languages.